### PR TITLE
Add `RichModelSummary` callback

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Follow up to #8929

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/rich_model_summary.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                table.add_column(column_name, justify="right", style="green")
+
+        rows = list(zip(*(arr[1] for arr in summary_data)))
+        for row in rows:
+            table.add_row(*row)
+
+        console.print(table)
+
+        # Formatting
+        s = "{:<{}}"
```
Comment (SeanNaren) [inline pytorch_lightning/callbacks/rich_model_summary.py:64]: Could you explain what this does 😅

File: pytorch_lightning/callbacks/rich_model_summary.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        model_size: float,
+    ) -> None:
+
+        console = Console()
+
+        table = Table(title="Model Summary")
+
+        table.add_column(" ")
+        table.add_column("Name", justify="left", style="cyan", no_wrap=True)
+        table.add_column("Type", style="magenta")
```
Comment (SeanNaren) [inline pytorch_lightning/callbacks/rich_model_summary.py:48]: Regarding our offline conversation, how could you see these styles being easier to modify? I was thinking some kinda config dict structure that the user can pass, or we just expose the colours to the `__init__`

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 3]
```diff
@@ -102,7 +111,12 @@ def _configure_model_summary_callback(self, weights_summary: Optional[str] = Non
                     f" but got {weights_summary}",
                 )
             max_depth = ModelSummaryMode.get_max_depth(weights_summary)
-            model_summary = ModelSummary(max_depth=max_depth)
+            if self.trainer._progress_bar_callback is not None and isinstance(
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/callback_connector.py:114]: @kaushikb11  don't we want to use rich summary when rich is installed automatically?

File: pytorch_lightning/callbacks/rich_model_summary.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.imports import _RICH_AVAILABLE
+from pytorch_lightning.utilities.model_summary import get_human_readable_count
+
+if _RICH_AVAILABLE:
+    from rich.console import Console
+    from rich.table import Table
+
+
+class RichModelSummary(ModelSummary):
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/rich_model_summary.py:26]: don't forget to add this file to the mypy list

Comment (awaelchli) [reply pytorch_lightning/callbacks/rich_model_summary.py:26]: and let's add a docstring here? that this requires rich and instructions how to install? also maybe point to rich progressbar.

File: pytorch_lightning/trainer/connectors/callback_connector.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/callback_connector.py:114]: @awaelchli we ran into one issue where we had an old rich version installed (8.x instead of 10.x), but the `_RICH_AVAILABLE` check defaulted to using the rich progress bar because the package was available. 

If we default to using Rich, let's make sure the package versioning requirements for Rich are also part of the availability check

File: pytorch_lightning/callbacks/rich_model_summary.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/rich_model_summary.py]: Removed!

File: pytorch_lightning/callbacks/rich_model_summary.py (hunk 2 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/rich_model_summary.py:82]: Yup, after your recent RichProgressBar PR gets merged, will add a config dataclass argument here to the signature as well!

File: pytorch_lightning/trainer/connectors/callback_connector.py (hunk 3 above)
Comment (kaushikb11) [reply pytorch_lightning/trainer/connectors/callback_connector.py:114]: There's still a discussion going on whether we should add it to Bolts or not. Would wait for that.

File: pytorch_lightning/callbacks/rich_model_summary.py (hunk 4 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/rich_model_summary.py:25]: Addressed!

File: pytorch_lightning/callbacks/rich_model_summary.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        pip install rich
+
+    .. code-block:: python
+
+        from pytorch_lightning import Trainer
+        from pytorch_lightning.callbacks import RichModelSummary
+
+        trainer = Trainer(callbacks=RichModelSummary())
+
+    You could also enable it using the :class:`~pytorch_lightning.callbacks.RichProgressBar`
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/rich_model_summary.py:43]: "it" here refers to the model summary above but I think you want to say that a user can also enable rich logging with the progress bar.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/callback_connector.py (hunk 3 above)
Comment (SeanNaren) [reply pytorch_lightning/trainer/connectors/callback_connector.py:114]: I'll make an issue for this, and we can enable this in a separate PR :)

**Review by Borda — APPROVED** [review]
